Self-learning enhancements and general cleanup (non-functional)

khalid-a-omar opened the pull request:
Remove unused and unnecessary variables

khalid-a-omar committed:
Code cleanup 1 (non-functional)

Remove unused variable

khalid-a-omar committed:
Code cleanup 2 (non-functional)

Remove unnecessary variables

khalid-a-omar committed:
Make sure experience file is saved in the same folder as the engine

Some GUI's do not set the current directory of the engine to the same directory of the engine itself. This causes the experience file to be created on the same folder as the GUI itself!

This patch aims to avoid this problem alltogether by specifying a full path to the experience file at runtime.

amchess reviewed (commented):
In misc.cpp, I propose to refactor DirectorySeperator in DirectorySeparator.

khalid-a-omar committed:
Fixing typo

khalid-a-omar commented:
> 
> 
> In misc.cpp, I propose to refactor DirectorySeperator in DirectorySeparator.

Agreed. Mine is a typo. Good catch
I fixed it

khalid-a-omar committed:
Self-learning enhancements

Many non-functional changes to enhance experience loading/saving, as follows:
- Save experience file when game is decided, i.e., before the game actually ends (Check is_game_decided() for more details)
- Fix few places where experience was getting modified even when the option "Read only learning" is set to true
- Fix for missing Q-learning logic when engine receives "quit" command
- Change experience saving logic to a failproof approach that guarentees we never lose the exiting experience data even if the saving gets interrupted for whatever reason (engine gets forcefully terminated, power failure, etc...)
- Change experience loading logic to speed up the loading by eliminating redundant load and re-creating of experience file unnecessarily
- Removed redundant code specially from the various experience loading/saving functions. Now all the loading takes place in loadExperienceFile()

khalid-a-omar committed:
Update tt.cpp

Removed unused function and unnecessary initializations

khalid-a-omar committed:
Update tt.cpp

Formatting!

khalid-a-omar committed:
Added missing condition to if statement

khalid-a-omar committed:
Typos!

amchess commented:
Very good work!
